File: 0881-loud-and-rich/0881-loud-and-rich.cpp
class Solution {
public:
    vector<int> loudAndRich(vector<vector<int>>& richer, vector<int>& quiet) {
        
        int  n = quiet.size();
        vector<int> adj[n];

        vector<int> indegree(n,0);

        vector<int> ans(n,0);

        queue<int> q;

        for(auto row : richer){

            adj[row[0]].push_back(row[1]);

            indegree[row[1]]++;
        }

        for(int i=0;i<n;i++){

            ans[i]= i;

            if(indegree[i]==0) q.push(i);
        }

        while(!q.empty()){

            int node = q.front();

            q.pop();

            for(auto next : adj[node]){

                if(quiet[ans[node]] < quiet[ans[next]] ){

                    ans[next] = ans[node];
                }

                indegree[next]--;

                if(indegree[next]==0) q.push(next);
            }
        }

        return ans;
    }
};

File: 0881-loud-and-rich/0881-loud-and-rich.java
class Solution {
    public int[] loudAndRich(int[][] richer, int[] quiet) {

        List<List<Integer>> adj = new ArrayList<>();

        int n = quiet.length;

        for(int i=0;i<n;i++) adj.add(new ArrayList<>());

        int[] indegree = new int[n];
        Queue<Integer> q = new LinkedList<>();
        int ans[] = new int[n];

        for(int[] row : richer){

            adj.get(row[0]).add(row[1]);
            indegree[row[1]]++;
            
        }

        for(int i=0;i<n;i++){

            ans[i] = i;

            if(indegree[i]==0) q.offer(i);
        }

        while(!q.isEmpty()){

            int node = q.poll();

            for(int next : adj.get(node)){

                if(quiet[ans[node]] < quiet[ans[next]]){

                    ans[next] = ans[node];
                }

                indegree[next]--;

                if(indegree[next]==0) q.offer(next);
            }
        }

        return ans; 
        
    }
}